### ISSUE TITLE
Can I use colmap as a library and triangulate 3d points from existing 2d points and matches?

### ISSUE THREAD

**Issue by jingedawang** [colmap/colmap#688]
**Is your feature request related to a problem? Please describe.**
I try to triangulate 3d points with existing 2d points and matches using colmap as a library. So I check the code in `RunPointTriangulator` method of colmap.cc. But this example use the SQLite database as input. It is not what I want, so I try to use `DatabaseCache` directly. But apparently, this class does not provide enough interface for construction without database. I am stuck here.

**Describe the solution you'd like**
I think `DatabaseCache` should provide one more interface like `AddCorrespondence`. Why not provide `AddCorrespondence` while `AddImage` and `AddCamera` are provided.

**Describe alternatives you've considered**
There is no alternatives now, unless I construct a database in advance. But it is meaningless and redundant.

**Additional context**
I need this feature because I must use my customized keypoints and descriptors instead of SIFT. And I want to accomplish the triangulation through memory, without the need for io read and write. I think it's a reasonable requirement.

**Comment by tsattler** [comment]
You should be able to do this with the functionality provided at the moment by importing the features into a database (see https://colmap.github.io/tutorial.html#feature-detection-and-extraction) and the importing custom matches (see https://colmap.github.io/tutorial.html#feature-matching-and-geometric-verification). The corresponding command line tools should be `feature_importer` and `matches_importer`.

**Comment by tsattler** [comment]
You can also take a look here https://github.com/tsattler/visuallocalizationbenchmark/tree/master/local_feature_evaluation for inspiration.

**Comment by jingedawang** [comment]
Thanks for your reply. I have already tried constructing database with command line, it does work. But I am pursuing a way git rid of I/O operation for the sake of performance. Since current COLMAP does not support it, I will try to implement it by myself. Will you consider to accept my modification afterwards?

**Comment by ahorn** [comment]
@jingedawang Out of curiosity, where you able to bypass the need of storing your feature descriptors in a database and use an in-memory data structure instead?

**Comment by jingedawang** [comment]
@ahorn I have my own data structure and database format, which is not the same as COLMAP's. When I use COLMAP as a library, what I want is just the mapping algorithm, not the whole processing pipeline.

**Comment by ahorn** [comment]
Thank you. In other words, you are currently not using COLMAP due to your aforementioned restrictions in the OP, correct?

**Comment by jingedawang** [comment]
Yes, I am not using it currently. I was trying to use it by modifying some source code as I mentioned above, but not accomplished till now and suspended recently.
